hashintel/hash
# Show placeholder on empty paragraphs on the editor

## 🌟 What is the purpose of this PR?

- Create a `Placeholder` component, and render it if editor is focused on an empty `Paragraph` block.
- On page load, focus on the first `Paragraph` of the page, if it's empty and the page only has one block (which is the `Paragraph`)


## ⚙️ Change on eslint config

Removed `DecorationSet` from the following two eslint rules. 
https://github.com/hashintel/hash/blob/21321047715d7a2e9087a2611fa1f3bcccbd787f/.eslintrc.json#L54-L61

To see why, check the code from `@types/prosemirror-view` below. We don't need to pass a generic to this, because it uses the argument's type as return type.

```ts
// https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/prosemirror-view/index.d.ts

/**
 * Create a set of decorations, using the structure of the given
 * document.
 */
static create<S extends Schema = any>(doc: ProsemirrorNode<S>, decorations: Decoration[]): DecorationSet<S>;
```

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1202621375435417/1202765662321678/f) _(internal)_
- [Design](https://www.figma.com/file/gydVGka9FjNEg9E2STwhi2/HASH-App?node-id=452%3A17402) _(internal)_

## ❓ How to test this?

2. Checkout the branch / view the deployment
3. Create a new page
4. It should focus into first paragraph block on it, and you should see the placeholder
5. Try refreshing the page, it should work again, by focusing into first paragraph block

## 📹 Demo

https://user-images.githubusercontent.com/39553853/185290741-c16748ee-667a-4bf4-b31e-588eccb80705.mov
